extensions/message-ids.md

```diff
@@ -0,0 +1,128 @@
+
+The message ID tag is a way for servers to enable these enhancements.
+
+## Architecture
+
+### Dependencies
+
+This specification doesn't define any capabilities of its own, but the `draft/message-tags` capability MUST be negotiated for servers wishing to use this tag.
+
+### Tags
+
+This specification adds the `draft/msgid` tag, which has a required value.
+
+Servers MAY attach this tag on any event. If this tag is being used, it SHOULD be attached to all `PRIVMSG` and `NOTICE` events.
+
+#### Tag value
+
+The tag value is chosen by the originating server and MUST be unique, meaning any other message transmitted on the entire network at any time MUST NOT share the same value.
+
+However, if a message is re-transmitted as-is, for example with the [`chathistory`](./batch/chathistory-3.3.html) batch type, the ID SHOULD be reused. As a result, clients MUST accept shared IDs.
+
+The tag value MUST be treated as a case sensitive opaque identifier. Clients MUST NOT use case folding or normalization when comparing IDs.
+
+### Relationships with other specifications
+
+In order to treat messages that refer to IDs consistently, clients need to know the IDs for their own messages as well. Servers that provide message IDs SHOULD also provide the [`echo-message`](./echo-message-3.2.html) capability.
+
+## Server implementation considerations
+
+This section is non-normative.
+
+### Choosing an ID format
+
+In order to guarantee sufficient uniqueness, message IDs can't be implemented as simple numeric counters that risk clashing with other servers on the network, or being reset if the server restarts.
+
+Some examples of appropriate IDs are:
+
+* An [RFC 4122 UUID](https://tools.ietf.org/html/rfc4122).
+* An ID generated by a shared external provider, used by all servers on the network.
+* A numeric counter, prefixed with a server ID, with the latest value saved externally to the IRCd on every issue, to allow for power failures.
+* A numeric counter, prefixed with a server ID as well as the server startup timestamp at a suitable precision.
+
+When using timestamps, make sure they're correctly synchronised using NTP or similar.
+
+Although clients are required to treat IDs as case sensitive opaque values, servers might still choose a case insensitive ID format internally.
+
+## Client implementation considerations
+
+This section is non-normative.
+
+Message IDs have no guarantee of being universally unique across different IRC networks, nor will they necessarily share the same format. There is also no requirement that numeric IDs increase monotonically. Don't attempt to correlate them beyond their scope and don't use them for message ordering.
```

Comment:
FWIW, it can prevent linking two smaller networks into one bigger network, if there is a chance that the same ID was generated by both networks in the past.

Related: during temporary netsplit, I assume the parts of network are still "single network".